### ISSUE TITLE
[bug]: segfault in main/rtp_engine.c:1489 after updating 20.8.1 -> 20.9.0

### ISSUE THREAD

**Issue by sircubbi** [asterisk/asterisk#822]
### Severity

Major

### Versions

20.9.0

### Components/Modules

rtp_engine

### Operating Environment

RockyLinux 9

### Frequency of Occurrence

Frequent

### Issue Description

After upgrading from 20.8.1 to 20.9.0 we encounter segfaults in rtp_engine.c and had to rollback to 20.8.1. I saw three segfaults so far within a 20 minute timeframe after a constant load of around 100-150 parallel calls.

It seems related to commit 3b4d2a65064eb7a49d1d765a80e8ffa602532d0a (Issue #699). Unfortunately I cannot provide a backtrace, since no core dump was captured. However I am able to pinpoint the segfault to line 1489 in `rtp_engine.c`, so hopefully that gives you a clue on where to look.

I am happy to provide more details if needed of course.

### Relevant log output

```shell
Kernel-Logfile:

Jul 25 10:13:10 sip kernel: asterisk[4004920]: segfault at 0 ip 0000000000535a91 sp 00007f857a188c40 error 4 in asterisk[438000+1d1000] likely on CPU 1 (core 0, socket 1)
Jul 25 10:13:32 sip kernel: asterisk[47139]: segfault at 0 ip 0000000000535a91 sp 00007f56243fcc40 error 4 in asterisk[438000+1d1000] likely on CPU 3 (core 0, socket 3)
Jul 25 10:31:17 sip kernel: asterisk[47329]: segfault at 0 ip 0000000000535a91 sp 00007f75e8c38c40 error 4 in asterisk[438000+1d1000] likely on CPU 3 (core 0, socket 3)
```

Determine the code in question:
```
# addr2line -e asterisk 0000000000535a91
/root/asterisk-src/asterisk-20.9.0/main/rtp_engine.c:1489
#
```

Diff between 20.8.1 and 20.9.0 for that code:
```
@@ -1457,6 +1485,10 @@
 
 	if (payload < AST_VECTOR_SIZE(&codecs->payload_mapping_tx)) {
 		type = AST_VECTOR_GET(&codecs->payload_mapping_tx, payload);
+		/* remove the preferred format if we are unsetting its container. */
+		if (ast_format_cmp(type->format, codecs->preferred_format) == AST_FORMAT_CMP_EQUAL) {
+			ao2_replace(codecs->preferred_format, NULL);
+		}
 		ao2_cleanup(type);
 		AST_VECTOR_REPLACE(&codecs->payload_mapping_tx, payload, NULL);
 	}
```

In conclusion the segfault happens at the call to `ast_format_cmp`.
```


### Asterisk Issue Guidelines

- [X] Yes, I have read the Asterisk Issue Guidelines

**Comment by jcolp** [comment]
What channel driver is in use?

**Comment by sircubbi** [comment]
> What channel driver is in use?

Sorry, I forgot to add. This is still with `chan_sip` I fear. (Migration to `res_pjsip` is ongoing, but still in progress 😔).

**Comment by jcolp** [comment]
Ah, that's the only thing that uses that function which is likely why noone has noticed this until now.
